cli/flox-rust-sdk/src/models/manifest.rs

```diff
@@ -3,11 +3,207 @@ use std::process::Command;
+    fn get_version(&self) -> i64 {
+        self.0
+            .get("version")
+            .and_then(Item::as_integer)
+            .unwrap_or(1)
+    }
+
+    /// Serde's error messages for _untagged_ enums are rather bad
+    /// and don't appear to become better any time soon:
+    /// - <https://github.com/serde-rs/serde/pull/1544>
+    /// - <https://github.com/serde-rs/serde/pull/2376>
+    ///
+    /// This function aims to provide the intermediate version matching on
+    /// the `version` field, and then deserializes the correct version
+    /// of the Manifest explicitly.
+    ///
+    /// <https://github.com/serde-rs/serde/pull/2525> will allow the use of integers
+    /// (i.e. versions) as enum tags, which will allow us to use `#[serde(tag = "version")]`
+    /// and avoid the [Version] field entirely, where the version field is not optional.
+    ///
+    /// Discussion: using a string field as the version tag `version: "1"` vs `version: 1`
+    /// could work today, but is still limited by the lack of an optional tag.
+    fn to_typed(&self) -> Result<TypedManifest, toml_edit::de::Error> {
+        match self.get_version() {
+            2 => Ok(TypedManifest::Catalog(toml_edit::de::from_document(
+                self.0.clone(),
+            )?)),
+            1 => Ok(TypedManifest::Pkgdb(toml_edit::de::from_document(
+                self.0.clone(),
+            )?)),
+            v => {
+                let msg = format!("unsupported manifest version: {v}");
+                Err(toml_edit::de::Error::custom(msg))
+            },
+        }
+    }
+}
+
+impl FromStr for RawManifest {
+    type Err = toml_edit::de::Error;
+
+    /// Parses a string to a `ManifestMut` and validates that it's a valid manifest
+    /// Validation is currently only checking the structure of the manifest,
+    /// not the precise contents.
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let doc = s.parse::<DocumentMut>()?;
+        let manifest = RawManifest(doc);
+        let _validate = manifest.to_typed()?;
+        Ok(manifest)
+    }
+}
+
+/// Represents the Manifest data schema for reading/processing of the manifest.
+/// Writing a [`TypedManifest`] will drop comments and formatting.
+/// Hence, this should only be used in cases where these can safely be severed.
+/// Edits to the user facing manifest.toml file should be made using [`RawManifest`] instead.
+#[derive(Debug, Serialize)]
+#[serde(untagged)]
+enum TypedManifest {
+    /// v2 manifest, processed by flox and resolved using the catalog service
+    Catalog(Box<TypedManifestCatalog>),
+    /// deprecated v1 manifest, processed entirely by `pkgdb`
+    Pkgdb(TypedManifestPkgdb),
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct TypedManifestPkgdb {
+    version: Option<Version<1>>,
+    vars: Option<toml::Table>,
+    hook: Option<toml::Table>,
+    #[serde(flatten)]
+    _toml: toml::Table,
+}
```

Comment:
A copy of the `Manifest` type he have already, but since renaming it tree wide would be a bit messy/distracting, I opted to deprecate `Manifest` and would aim to replace it with `TypedManifest` where possible.